<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Reading in and filtering data • MIPanalyzer</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Reading in and filtering data">
<meta property="og:description" content="MIPanalyzer">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MIPanalyzer</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/filtering_data.html">Filtering data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Reading in and filtering data</h1>
                        <h4 data-toc-skip class="author">Bob Verity</h4>
            
            <h4 data-toc-skip class="date">Last updated: 11 Dec
2023</h4>
      
      
      <div class="hidden name"><code>filtering_data.Rmd</code></div>

    </div>

    
    
<p>This tutorial covers:</p>
<ul>
<li>Reading in data from a <code>vcf</code> file into MIPanalyzer
format</li>
<li>Filtering based on coverage</li>
<li>Filtering based on other criteria</li>
</ul>
<div class="section level2">
<h2 id="reading-in-data-and-understanding-the-mipanalyzer_biallelic-format">Reading in data and understanding the
<code>mipanalyzer_biallelic</code> format<a class="anchor" aria-label="anchor" href="#reading-in-data-and-understanding-the-mipanalyzer_biallelic-format"></a>
</h2>
<p>Before doing anything, we will need to load some additional
packages:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load some packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://here.r-lib.org/" class="external-link">here</a></span><span class="op">)</span></span></code></pre></div>
<p>We will be working from a <code>vcf</code> file stored in the
<code>inst/extdata</code> folder of this package. Our first decision
point is whether we want to read this <code>vcf</code> file in allowing
for more than two alleles at any locus, or forcing data to bi-allelic.
Many analyses assume bi-allelic SNPs, which also makes the data format
more efficient as we only need to encode the read <code>counts</code>
and <code>coverage</code> as single numbers, rather than storing
multiple <code>counts</code> for each allele. Here we will assume
bi-allelic data, but there will be a section below on the multi-allelic
data format.</p>
<p>The <code><a href="../reference/vcf_to_mipanalyzer_biallelic.html">vcf_to_mipanalyzer_biallelic()</a></code> can read in file and
convert it to our custom format. It accepts a path to a <code>vcf</code>
file, or alternatively you can read the <code>vcf</code> in yourself and
pass this to the function:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># define the path to your .vcf file</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>vcf_path <span class="ot">&lt;-</span> <span class="fu">here</span>(<span class="st">"inst/extdata"</span>, <span class="st">"DRC_pf3k_filtered.vcf"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co"># load as mipanalyzer_biallelic object</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>dat_biallelic <span class="ot">&lt;-</span> <span class="fu">vcf_to_mipanalyzer_biallelic</span>(<span class="at">file =</span> vcf_path)</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="co">#&gt; Scanning file to determine attributes.</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="co">#&gt; File attributes:</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="co">#&gt;   meta lines: 135</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="co">#&gt;   header_line: 136</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="co">#&gt;   variant count: 2667</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="co">#&gt;   column count: 122</span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a>Meta line <span class="dv">135</span> read in.</span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a><span class="co">#&gt; All meta lines processed.</span></span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a><span class="co">#&gt; gt matrix initialized.</span></span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a><span class="co">#&gt; Character matrix gt created.</span></span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a><span class="co">#&gt;   Character matrix gt rows: 2667</span></span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a><span class="co">#&gt;   Character matrix gt cols: 122</span></span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a><span class="co">#&gt;   skip: 0</span></span>
<span id="cb2-20"><a href="#cb2-20" tabindex="-1"></a><span class="co">#&gt;   nrows: 2667</span></span>
<span id="cb2-21"><a href="#cb2-21" tabindex="-1"></a><span class="co">#&gt;   row_num: 0</span></span>
<span id="cb2-22"><a href="#cb2-22" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-23"><a href="#cb2-23" tabindex="-1"></a>Processed variant <span class="dv">1000</span></span>
<span id="cb2-24"><a href="#cb2-24" tabindex="-1"></a>Processed variant <span class="dv">2000</span></span>
<span id="cb2-25"><a href="#cb2-25" tabindex="-1"></a>Processed variant<span class="sc">:</span> <span class="dv">2667</span></span>
<span id="cb2-26"><a href="#cb2-26" tabindex="-1"></a><span class="co">#&gt; All variants processed</span></span>
<span id="cb2-27"><a href="#cb2-27" tabindex="-1"></a><span class="co">#&gt; Processing</span></span>
<span id="cb2-28"><a href="#cb2-28" tabindex="-1"></a><span class="co">#&gt; Done</span></span>
<span id="cb2-29"><a href="#cb2-29" tabindex="-1"></a></span>
<span id="cb2-30"><a href="#cb2-30" tabindex="-1"></a><span class="co"># take a peek</span></span>
<span id="cb2-31"><a href="#cb2-31" tabindex="-1"></a>dat_biallelic</span>
<span id="cb2-32"><a href="#cb2-32" tabindex="-1"></a><span class="co">#&gt; samples = 113</span></span>
<span id="cb2-33"><a href="#cb2-33" tabindex="-1"></a><span class="co">#&gt; loci = 2667</span></span></code></pre></div>
<p>We can see that we are working with 113 samples and 2667 loci. This
dataset is actually filtered down from <a href="https://www.malariagen.net/parasite/pf3k" class="external-link">Pf3k</a> whole genome
data using samples from the Democratic Republic of the Congo (DRC).
Although this object is its own custom class, it is essentially a
list:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">dat_biallelic</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "mipanalyzer_biallelic"</span></span>
<span></span>
<span><span class="co"># look at sub-object names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">dat_biallelic</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "coverage"       "counts"         "samples"        "loci"          </span></span>
<span><span class="co">#&gt; [5] "filter_history" "vcfmeta"</span></span></code></pre></div>
<p>The <code>coverage</code> and <code>counts</code> objects are our
main genetic data:</p>
<ul>
<li>
<code>coverage</code> is the “DP” element of the <code>vcf</code>,
which gives the combined read depth at this position.</li>
<li>
<code>counts</code> is the <em>first</em> value in the “AD” element
of the <code>vcf</code>, which gives the read depth of the first allele.
We assume that this is the reference (REF) allele by convention.</li>
</ul>
<p>Both of these objects are matrices, with samples in rows and loci in
columns:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">dat_biallelic</span><span class="op">$</span><span class="va">coverage</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]  113 2667</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">dat_biallelic</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]  113 2667</span></span></code></pre></div>
<p>The next object is the <code>samples</code> element. This is a
data.frame containing the sample names from the <code>vcf</code>. You
can add whatever columns you like to this data.frame, making it easy to
keep track of meta-data alongside genetic data.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">dat_biallelic</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 113   1</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dat_biallelic</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span></span>
<span><span class="co">#&gt;   SAMPLE_ID</span></span>
<span><span class="co">#&gt; 1  QG0182-C</span></span>
<span><span class="co">#&gt; 2  QG0183-C</span></span>
<span><span class="co">#&gt; 3  QG0184-C</span></span>
<span><span class="co">#&gt; 4  QG0185-C</span></span>
<span><span class="co">#&gt; 5  QG0186-C</span></span>
<span><span class="co">#&gt; 6  QG0187-C</span></span></code></pre></div>
<p>Next up is the <code>loci</code> element. This is a data.frame
containing all the information in the “fix” element of the
<code>vcf</code>:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">dat_biallelic</span><span class="op">$</span><span class="va">loci</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 2667    8</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dat_biallelic</span><span class="op">$</span><span class="va">loci</span><span class="op">)</span></span>
<span><span class="co">#&gt;         CHROM   POS   ID REF ALT    QUAL FILTER</span></span>
<span><span class="co">#&gt; 1 Pf3D7_01_v3 95518 &lt;NA&gt;   G   T 7926880   PASS</span></span>
<span><span class="co">#&gt; 2 Pf3D7_01_v3 95632 &lt;NA&gt;   G   A 6345960   PASS</span></span>
<span><span class="co">#&gt; 3 Pf3D7_01_v3 95641 &lt;NA&gt;   G   A 6106740   PASS</span></span>
<span><span class="co">#&gt; 4 Pf3D7_01_v3 95680 &lt;NA&gt;   A   G 5786050   PASS</span></span>
<span><span class="co">#&gt; 5 Pf3D7_01_v3 95685 &lt;NA&gt;   G   T 5966510   PASS</span></span>
<span><span class="co">#&gt; 6 Pf3D7_01_v3 95686 &lt;NA&gt;   A   C 5924990   PASS</span></span>
<span><span class="co">#&gt;                                                                            INFO</span></span>
<span><span class="co">#&gt; 1 AC=154;AF=0.634;AN=226;DP=404727;RegionType=Core;VQSLOD=10.78;VariantType=SNP</span></span>
<span><span class="co">#&gt; 2  AC=85;AF=0.372;AN=224;DP=351431;RegionType=Core;VQSLOD=11.47;VariantType=SNP</span></span>
<span><span class="co">#&gt; 3  AC=85;AF=0.374;AN=222;DP=345575;RegionType=Core;VQSLOD=11.95;VariantType=SNP</span></span>
<span><span class="co">#&gt; 4  AC=84;AF=0.365;AN=222;DP=329961;RegionType=Core;VQSLOD=11.99;VariantType=SNP</span></span>
<span><span class="co">#&gt; 5  AC=84;AF=0.365;AN=222;DP=329708;RegionType=Core;VQSLOD=12.02;VariantType=SNP</span></span>
<span><span class="co">#&gt; 6  AC=84;AF=0.367;AN=222;DP=330180;RegionType=Core;VQSLOD=12.03;VariantType=SNP</span></span></code></pre></div>
<p>Next is the <code>filter_history</code> element. This contains a
running record of all filters that are applied to the data via the
MIPanalyzer package. Currently, we have applied no filters and so what
we see are the raw dimensions of the data. We will return to this object
later:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat_biallelic</span><span class="op">$</span><span class="va">filter_history</span></span>
<span><span class="co">#&gt;   description samples loci n_missing prop_missing function_call</span></span>
<span><span class="co">#&gt; 1    raw data     113 2667       478  0.001586085            NA</span></span></code></pre></div>
<p>Finally, the <code>vcfmeta</code> element contains anything in the
“meta” slot of the <code>vcf</code>. There may be a wide range of
different information here, which is stored as a list. Here are just the
first three elements in our example data:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat_biallelic</span><span class="op">$</span><span class="va">vcfmeta</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "##fileformat=VCFv4.2"                                                                          </span></span>
<span><span class="co">#&gt; [2] "##FILTER=&lt;ID=PASS,Description=\"All filters passed\"&gt;"                                         </span></span>
<span><span class="co">#&gt; [3] "##ALT=&lt;ID=NON_REF,Description=\"Represents any possible alternative allele at this location\"&gt;"</span></span></code></pre></div>
<p>In summary, the <code>mipanalyzer_biallelic</code> class of data is
just a re-coding of the <code>vcf</code>. It exposes the
<code>coverage</code> and <code>counts</code> data, while also nice
features in terms of keeping track of meta-data and filters.</p>
</div>
<div class="section level2">
<h2 id="filtering-data-based-on-coverage">Filtering data based on coverage<a class="anchor" aria-label="anchor" href="#filtering-data-based-on-coverage"></a>
</h2>
<p>We very often want to focus our attention on sites that have good
coverage. Low coverage sites lead to uncertain estimates of
within-sample allele frequencies, and may also be a sign of sequencing
problems.</p>
<p>We could choose to filter by sample, or by locus. Loci that have
systematically low coverage over many samples may indicate issues with
sequencing. We can use the <code><a href="../reference/explore_filter_coverage_loci.html">explore_filter_coverage_loci()</a></code>
to explore this issue via the following steps:</p>
<ul>
<li>Set a <code>min_coverage</code> threshold. Any <code>coverage</code>
value less than this threshold we define as <em>low coverage</em>.</li>
<li>Count the percentage of samples that are <em>low coverage</em>. Do
this for every locus.</li>
<li>Plot a histogram of this percentage</li>
<li>Set the <code>max_low_coverage</code> that we will accept as a
percentage. Work out what how many samples fall below this level.</li>
</ul>
<p>This gives an indication of how many samples we will lose if we apply
this filter, and produces the following plot:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat_biallelic</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/explore_filter_coverage_loci.html">explore_filter_coverage_loci</a></span><span class="op">(</span>min_coverage <span class="op">=</span> <span class="fl">10</span>, max_low_coverage <span class="op">=</span> <span class="fl">25</span><span class="op">)</span></span></code></pre></div>
<p><img src="filtering_data_files/figure-html/unnamed-chunk-11-1.png" width="576"></p>
<p>We can see that we will lose 3.22% of samples if we apply this
filter, which is acceptable. We can therefore go ahead and apply the
filter:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat_biallelic</span> <span class="op">&lt;-</span> <span class="va">dat_biallelic</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/filter_coverage_loci.html">filter_coverage_loci</a></span><span class="op">(</span>min_coverage <span class="op">=</span> <span class="fl">10</span>, max_low_coverage <span class="op">=</span> <span class="fl">25</span><span class="op">)</span></span></code></pre></div>
<p>Next, we can perform the same task but looking at it from a sample
perspective. This involves the following steps:</p>
<ul>
<li>Set a <code>min_coverage</code> threshold. Any <code>coverage</code>
value less than this threshold we define as <em>low coverage</em>.</li>
<li>Count the percentage of loci that are <em>low coverage</em>. Do this
for every sample.</li>
<li>Plot a histogram of this percentage</li>
<li>Set the <code>max_low_coverage</code> that we will accept as a
percentage. Work out what how many loci fall below this level.</li>
</ul>
<p>We obtain the following plot:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat_biallelic</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/explore_filter_coverage_samples.html">explore_filter_coverage_samples</a></span><span class="op">(</span>min_coverage <span class="op">=</span> <span class="fl">10</span>, max_low_coverage <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p><img src="filtering_data_files/figure-html/unnamed-chunk-13-1.png" width="576"></p>
<p>Note that from the shape of the histogram we have good coverage for
many samples, and then a few samples that perform poorly with low
coverage over many loci. We therefore set quite a stringent threshold of
<code>max_low_coverage = 10</code> to exclude these samples. We can see
that we will lose 20.35% of samples, which is a fairly stringent filter,
but will leave us with high quality samples.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat_biallelic</span> <span class="op">&lt;-</span> <span class="va">dat_biallelic</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/filter_coverage_samples.html">filter_coverage_samples</a></span><span class="op">(</span>min_coverage <span class="op">=</span> <span class="fl">10</span>, max_low_coverage <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p>Every time we apply one of these filters, the change to data
dimension is stored in the <code>filter_history</code> element along
with the function call itself:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat_biallelic</span><span class="op">$</span><span class="va">filter_history</span></span>
<span><span class="co">#&gt;                        description samples loci n_missing prop_missing</span></span>
<span><span class="co">#&gt; 1                         raw data     113 2667       478 0.0015860849</span></span>
<span><span class="co">#&gt; 2    filter loci based on coverage     113 2581       267 0.0009154715</span></span>
<span><span class="co">#&gt; 3 filter samples based on coverage      91 2581       146 0.0006216178</span></span>
<span><span class="co">#&gt;                                                                              function_call</span></span>
<span><span class="co">#&gt; 1                                                                                     &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 2        filter_coverage_loci(x = dat_biallelic, min_coverage = 10, max_low_coverage = 25)</span></span>
<span><span class="co">#&gt; 3 filter_coverage_samples(x = dat_biallelic, min_coverage = 10,     max_low_coverage = 10)</span></span></code></pre></div>
<p>We can see that we are left with 91 samples, down from 113 initially,
and 2581 loci, down from 2667. Our proportion of missing data has
dropped considerably through these filters.</p>
</div>
<div class="section level2">
<h2 id="filtering-based-on-other-criteria">Filtering based on other criteria<a class="anchor" aria-label="anchor" href="#filtering-based-on-other-criteria"></a>
</h2>
<p>Above, we filtered based on coverage, but we may also want to exclude
loci/samples for other reasons. The <code><a href="../reference/filter_overcounts.html">filter_overcounts()</a></code>
function replaces any cell where the <code>count</code> exceeds the
<code>coverage</code> with <code>NA</code>. It should technically not be
possible for <code>count</code> to exceed <code>coverage</code> due to
the way they are defined, but this can come about due to upstream
issues. Note that we are not actually dropping any samples or loci via
this filter, we are just recoding information as missing.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># deal with count&gt;coverage issues</span></span>
<span><span class="va">dat_biallelic</span> <span class="op">&lt;-</span> <span class="va">dat_biallelic</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/filter_overcounts.html">filter_overcounts</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Now that we have discarded some samples, we may be left with some
loci that are no longer variable in the samples that remain. We probably
want to discard these as they do not contain useful information. We can
do this via the <code><a href="../reference/filter_loci_invariant.html">filter_loci_invariant()</a></code> function:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># drop loci that are no longer variable</span></span>
<span><span class="va">dat_biallelic</span> <span class="op">&lt;-</span> <span class="va">dat_biallelic</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/filter_loci_invariant.html">filter_loci_invariant</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Again, we can look at our filter history to see how these changes
have impacted our data dimensions:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat_biallelic</span><span class="op">$</span><span class="va">filter_history</span></span>
<span><span class="co">#&gt;                           description samples loci n_missing prop_missing</span></span>
<span><span class="co">#&gt; 1                            raw data     113 2667       478 0.0015860849</span></span>
<span><span class="co">#&gt; 2       filter loci based on coverage     113 2581       267 0.0009154715</span></span>
<span><span class="co">#&gt; 3    filter samples based on coverage      91 2581       146 0.0006216178</span></span>
<span><span class="co">#&gt; 4          replace overcounts with NA      91 2581       146 0.0006216178</span></span>
<span><span class="co">#&gt; 5 filter loci to drop invariant sites      91 2366       146 0.0006781047</span></span>
<span><span class="co">#&gt;                                                                              function_call</span></span>
<span><span class="co">#&gt; 1                                                                                     &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 2        filter_coverage_loci(x = dat_biallelic, min_coverage = 10, max_low_coverage = 25)</span></span>
<span><span class="co">#&gt; 3 filter_coverage_samples(x = dat_biallelic, min_coverage = 10,     max_low_coverage = 10)</span></span>
<span><span class="co">#&gt; 4                                                     filter_overcounts(x = dat_biallelic)</span></span>
<span><span class="co">#&gt; 5                 filter_loci(x = x, locus_filter = !invariant, description = description)</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Bob Verity.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
