<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Comparing genetic distances • MIPanalyzer</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Comparing genetic distances">
<meta property="og:description" content="MIPanalyzer">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MIPanalyzer</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/filtering_data.html">Filtering data</a>
    </li>
    <li>
      <a href="../articles/structure_analysis.html">Structure analysis</a>
    </li>
    <li>
      <a href="../articles/genetic_distances.html">Comparing genetic distances</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Comparing genetic distances</h1>
                        <h4 data-toc-skip class="author">Bob Verity</h4>
            
            <h4 data-toc-skip class="date">Last updated: 12 Dec
2023</h4>
      
      
      <div class="hidden name"><code>genetic_distances.Rmd</code></div>

    </div>

    
    
<p>This tutorial covers:</p>
<ul>
<li>Estimating relatedness between samples via maximum likelihood.</li>
<li>Comparing different genetic distance metrics.</li>
</ul>
<div class="section level2">
<h2 id="estimating-relatedness-between-samples">Estimating relatedness between samples<a class="anchor" aria-label="anchor" href="#estimating-relatedness-between-samples"></a>
</h2>
<p>Before doing anything, we will need to load some additional
packages:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load some packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://here.r-lib.org/" class="external-link">here</a></span><span class="op">)</span></span></code></pre></div>
<p>We will use a dataset of bi-allelic data already in the
<code>mipanalyzer_biallelic</code> format and filtered down to the
samples of interest (see earlier tutorial). We can load this example
data as follows:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># read mipanalyzer_biallelic from path</span></span>
<span><span class="va">dat_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"inst/extdata"</span>, <span class="st">"dat_biallelic.rds"</span><span class="op">)</span></span>
<span><span class="va">dat_biallelic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">dat_path</span><span class="op">)</span></span></code></pre></div>
<p>We want to estimate how related each sample is to each other as a
proportion between 0 and 1. This is not the same thing as simply
comparing genotypes because a samples may be have identical information
at a given locus (identity in state, IBS) without sharing a recent
common ancestor (identity by descent, IBD). In order to work out IBD, we
have to somehow account for the chance probability of being the
same.</p>
<p>One approach that was used <a href="https://doi.org/10.1038/s41467-020-15779-8" class="external-link">in this paper</a> is
to use the population-level allele frequencies (PLAFs) as our model for
how likely a sample is to carry the REF vs. ALT allele. Let <span class="math inline">\(p\)</span> be the PLAF of the REF allele at a
given locus, and <span class="math inline">\(q=1-p\)</span> be the PLAF
of the ALT allele. We can write down the probability of two samples
being identical for the REF allele (<span class="math inline">\(\text{Homo REF}\)</span>), identical for the ALT
allele (<span class="math inline">\(\text{Homo ALT}\)</span>), or
non-identical (<span class="math inline">\(\text{Het}\)</span>) as a
function of these allele frequencies and the inbreeding coefficient,
<span class="math inline">\(f\)</span>:</p>
<p><span class="math inline">\(Pr(\text{Homo REF}) = fp +
(1-f)p^2\)</span></p>
<p><span class="math inline">\(Pr(\text{Homo ALT}) = fq +
(1-f)q^2\)</span></p>
<p><span class="math inline">\(Pr(\text{Het}) = (1 - f)2pq\)</span></p>
<p>Taking the product of these probabilities over all loci, we arrive at
a likelihood in terms of <span class="math inline">\(f\)</span>. We can
then find the maximum likelihood estimate (MLE) of <span class="math inline">\(f\)</span>. Note that this approach assumes that
every locus is an independent draw from the PLAFs and that <span class="math inline">\(f\)</span> is independent across loci, both of
which are untrue. A more thorough approach is presented in tools like <a href="https://doi.org/10.1186/s12936-018-2349-7" class="external-link">hmmIBD</a> or <a href="https://doi.org/10.1371/journal.pgen.1007279" class="external-link">isoRelate</a>.
Nevertheless, this is a quick and easy way of getting an impression of
relatedness between samples.</p>
<p>The function <code><a href="../reference/inbreeding_mle.html">inbreeding_mle()</a></code> performs all of the steps
above on all pairwise samples:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mle_inbreeding</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/inbreeding_mle.html">inbreeding_mle</a></span><span class="op">(</span><span class="va">dat_biallelic</span>, report_progress <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mle_inbreeding</span><span class="op">$</span><span class="va">mle</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;      [,1] [,2] [,3] [,4] [,5]</span></span>
<span><span class="co">#&gt; [1,]   NA    0    0    0    0</span></span>
<span><span class="co">#&gt; [2,]   NA   NA    0    0    0</span></span>
<span><span class="co">#&gt; [3,]   NA   NA   NA    0    0</span></span>
<span><span class="co">#&gt; [4,]   NA   NA   NA   NA    0</span></span>
<span><span class="co">#&gt; [5,]   NA   NA   NA   NA   NA</span></span></code></pre></div>
<p>We can see that for the first 5 samples our best estimate is that
these samples are completely unrelated. We can convert this into a
genetic distance by taking one minus the relatedness, which results less
related samples having a greater distance between them. We can then
visualise the matrix of distances between all samples:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># convert to a distance</span></span>
<span><span class="va">dist_inbreeding</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="op">-</span> <span class="va">mle_inbreeding</span><span class="op">$</span><span class="va">mle</span></span>
<span></span>
<span><span class="co"># plot all pairwise distances between samples</span></span>
<span><span class="fu"><a href="../reference/plot_distance.html">plot_distance</a></span><span class="op">(</span><span class="va">dist_inbreeding</span>, col_pal <span class="op">=</span> <span class="st">"magma"</span><span class="op">)</span></span></code></pre></div>
<p><img src="genetic_distances_files/figure-html/unnamed-chunk-6-1.png" width="480"></p>
<p>There are a few pairs of samples that jump out as having low
distances (high relatedness). We can identify these samples:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># find the highly related pairs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">dist_inbreeding</span> <span class="op">&lt;=</span> <span class="fl">0.1</span>, arr.ind <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt;      row col</span></span>
<span><span class="co">#&gt; [1,]  53  54</span></span>
<span><span class="co">#&gt; [2,]  49  58</span></span>
<span><span class="co">#&gt; [3,]  69  71</span></span></code></pre></div>
<p>We can see that samples 53 and 54 appear to be highly related. It is
interesting to look at the within-sample allele frequencies (WSAFs) of
all samples from 51 to 55:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot WSAFs of samples 51 through 55</span></span>
<span><span class="va">wsaf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_wsaf.html">get_wsaf</a></span><span class="op">(</span><span class="va">dat_biallelic</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_wsaf.html">plot_wsaf</a></span><span class="op">(</span><span class="va">wsaf</span><span class="op">[</span><span class="fl">51</span><span class="op">:</span><span class="fl">55</span>,<span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="genetic_distances_files/figure-html/unnamed-chunk-8-1.png" width="672"></p>
<p>Most samples have clearly different patterns of WSAFs, but we can see
that samples 53 and 54 (the 3rd and 4th rows of this plot) have almost
identical WSAFs everywhere. This is a sanity check of what the MLE
inbreeding value is telling us - that these samples are more similar
than we would expect by chance. It is very likely that these are
identical (clonal) strains in two individuals, which may imply a short
transmission chain between them.</p>
</div>
<div class="section level2">
<h2 id="comparing-different-distance-metrics">Comparing different distance metrics<a class="anchor" aria-label="anchor" href="#comparing-different-distance-metrics"></a>
</h2>
<p>Several different methods for calculating genetic distances are
implemented inside <code>MIPanalyzer</code>. Let’s calculate a few of
them and compare results:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dist_genomic</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_genomic_distance.html">get_genomic_distance</a></span><span class="op">(</span><span class="va">dat_biallelic</span>, report_progress <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">dist_IBS</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_IBS_distance.html">get_IBS_distance</a></span><span class="op">(</span><span class="va">dat_biallelic</span>, report_progress <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">dist_mixture</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="op">-</span> <span class="fu"><a href="../reference/get_IB_mixture.html">get_IB_mixture</a></span><span class="op">(</span><span class="va">dat_biallelic</span>, report_progress <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">upper</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lower.tri.html" class="external-link">upper.tri</a></span><span class="op">(</span><span class="va">dist_genomic</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>MLE_inbreeding <span class="op">=</span> <span class="va">dist_inbreeding</span><span class="op">[</span><span class="va">upper</span><span class="op">]</span>,</span>
<span>           genomic <span class="op">=</span> <span class="va">dist_genomic</span><span class="op">[</span><span class="va">upper</span><span class="op">]</span>,</span>
<span>           IBS <span class="op">=</span> <span class="va">dist_IBS</span><span class="op">[</span><span class="va">upper</span><span class="op">]</span>,</span>
<span>           mixture <span class="op">=</span> <span class="va">dist_mixture</span><span class="op">[</span><span class="va">upper</span><span class="op">]</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="genetic_distances_files/figure-html/unnamed-chunk-9-1.png" width="672"></p>
<p>We can see that the “genomic” distance and the “MLE inbreeding”
distance agree quite closely. Also the “IBS” distance and the “mixture”
distance agree closely. However, the two sets of distances do not agree
with each other for certain key samples. These are the three samples
that we identified above as being closely related. This is because the
“genomic” and “MLE inbreeding” metrics are specifically designed to
identify outliers in terms of genetic similarities, whereas the other
two metrics are more crude estimates of genetic similarity.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Bob Verity.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
