<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analysis of population structure â€¢ MIPanalyzer</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analysis of population structure">
<meta property="og:description" content="MIPanalyzer">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MIPanalyzer</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/filtering_data.html">Filtering data</a>
    </li>
    <li>
      <a href="../articles/structure_analysis.html">Structure analysis</a>
    </li>
    <li>
      <a href="../articles/genetic_distances.html">Comparing genetic distances</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Analysis of population structure</h1>
                        <h4 data-toc-skip class="author">Bob Verity</h4>
            
            <h4 data-toc-skip class="date">Last updated: 12 Dec
2023</h4>
      
      
      <div class="hidden name"><code>structure_analysis.Rmd</code></div>

    </div>

    
    
<p>This tutorial covers:</p>
<ul>
<li>Principal Component Analysis (PCA) on within-sample allele
frequencies</li>
<li>Principal Coordinates Analsis (PCoA) on pairwise genetic
distances</li>
</ul>
<div class="section level2">
<h2 id="principal-component-analysis-pca">Principal Component Analysis (PCA)<a class="anchor" aria-label="anchor" href="#principal-component-analysis-pca"></a>
</h2>
<p>Before doing anything, we will need to load some additional
packages:</p>
<p>We will use a dataset of bi-allelic data already in the
<code>mipanalyzer_biallelic</code> format and filtered down to the
samples of interest (see previous tutorial). We can load this example
data as follows:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load data from within this package</span></span>
<span><span class="va">dat_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"inst/extdata"</span>, <span class="st">"dat_biallelic.rds"</span><span class="op">)</span></span>
<span><span class="va">dat_biallelic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">dat_path</span><span class="op">)</span></span></code></pre></div>
<p>The within-sample allele frequency (WSAF) can be simply calculated as
<code>count</code> divided by <code>coverage</code>. This is carried out
by the <code><a href="../reference/get_wsaf.html">get_wsaf()</a></code> function. Note that this function
automatically imputes missing values as the mean of all non-missing
values at each locus. This results in a complete matrix of the WSAF, but
we should keep in mind that some of this is imputed data.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># calculate within-sample allele frequencies</span></span>
<span><span class="va">wsaf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_wsaf.html">get_wsaf</a></span><span class="op">(</span><span class="va">dat_biallelic</span>, <span class="op">)</span></span></code></pre></div>
<p>It can be useful to produce a simple matrix plot of WSAFs:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_wsaf.html">plot_wsaf</a></span><span class="op">(</span><span class="va">wsaf</span><span class="op">)</span></span></code></pre></div>
<p><img src="structure_analysis_files/figure-html/unnamed-chunk-6-1.png" width="576"></p>
<p>We will carry out PCA using these values.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># calculate PCA</span></span>
<span><span class="va">pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pca_wsaf.html">pca_wsaf</a></span><span class="op">(</span><span class="va">wsaf</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot first 3 components</span></span>
<span><span class="fu"><a href="../reference/plot_pca.html">plot_pca</a></span><span class="op">(</span><span class="va">pca</span>, num_components <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://github.com/mrc-ide/MIPanalyzer/raw/master/vignettes/PCA_scatter.png" style="width: 80%"></p>
<p>This is a static image of the plot you will see - you will be able to
rotate and zoom it. We do not see any clear signal of clustering in this
plot, which may indicate that allele frequencies are relatively
homogeneous throughout these samples.</p>
<p>We may also want to know how much variance is explained by each
component:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot percentage variance explained</span></span>
<span><span class="fu"><a href="../reference/plot_pca_variance.html">plot_pca_variance</a></span><span class="op">(</span><span class="va">pca</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://github.com/mrc-ide/MIPanalyzer/raw/master/vignettes/PCA_variance.png" style="width: 60%"></p>
<p>Only a small percentage of the variance is explained by each
component, which again indicates that there is weak structure in the
data.</p>
<p>Finally, we may be interested in which loci are contributing the most
to this PCA variation. This can be obtained through the loading values
of a PCA. We can plot these values as a function of genomic position
using the <code><a href="../reference/plot_pca_contribution.html">plot_pca_contribution()</a></code> function. Note, this
first requires us to extract the CHROM of each locus in numerical
format:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get CHROM in numeric format for each locus</span></span>
<span><span class="va">chrom_numeric</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"_"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>, <span class="va">dat_biallelic</span><span class="op">$</span><span class="va">loci</span><span class="op">$</span><span class="va">CHROM</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot loading values</span></span>
<span><span class="fu"><a href="../reference/plot_pca_contribution.html">plot_pca_contribution</a></span><span class="op">(</span><span class="va">pca</span>, component <span class="op">=</span> <span class="fl">1</span>, chrom <span class="op">=</span> <span class="va">chrom_numeric</span>, pos <span class="op">=</span> <span class="va">dat_biallelic</span><span class="op">$</span><span class="va">loci</span><span class="op">$</span><span class="va">POS</span><span class="op">)</span></span></code></pre></div>
<p><img src="structure_analysis_files/figure-html/unnamed-chunk-9-1.png" width="672"></p>
<p>Interestingly, it appears that a large proportion of the variance in
WSAFs between samples is being driven by a few loci on chromosome 10.
This would be a good candidate area to eplore in terms of what could be
driving this signal.</p>
</div>
<div class="section level2">
<h2 id="principal-coordinates-analysis-pcoa-from-genetic-distances">Principal Coordinates Analysis (PCoA) from genetic distances<a class="anchor" aria-label="anchor" href="#principal-coordinates-analysis-pcoa-from-genetic-distances"></a>
</h2>
<p>An alternative way of exploring this data is to use principal
coordinates analysis (PCoA). Unlike PCA, this works with an input matrix
of distances, allowing us to be flexible with how we calculate these
distances. We will use a <a href="https://doi.org/10.7554/eLife.08714" class="external-link">genomic distance measure</a>
described by the MalariaGen community that allows for distances to be
calculated even between mixed infections.</p>
<p>We start by calculating the matrix of distances:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get genomic distance</span></span>
<span><span class="va">gdist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_genomic_distance.html">get_genomic_distance</a></span><span class="op">(</span><span class="va">dat_biallelic</span>, report_progress <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>The output is an upper triangle matrix of distances between all
samples:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gdist</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;      [,1]       [,2]       [,3]       [,4]       [,5]</span></span>
<span><span class="co">#&gt; [1,]   NA 0.08794438 0.09036943 0.09129772 0.08730510</span></span>
<span><span class="co">#&gt; [2,]   NA         NA 0.08587223 0.08747173 0.08538641</span></span>
<span><span class="co">#&gt; [3,]   NA         NA         NA 0.08921462 0.08786440</span></span>
<span><span class="co">#&gt; [4,]   NA         NA         NA         NA 0.08578565</span></span>
<span><span class="co">#&gt; [5,]   NA         NA         NA         NA         NA</span></span></code></pre></div>
<p>We can use the function <code><a href="../reference/pcoa_genomic_distance.html">pcoa_genomic_distance()</a></code> to
perform PCoA, and the function <code><a href="../reference/plot_pcoa.html">plot_pcoa()</a></code> to plot the
result:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># perform PCoA</span></span>
<span><span class="va">pcoa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pcoa_genomic_distance.html">pcoa_genomic_distance</a></span><span class="op">(</span><span class="va">gdist</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># scatterplot</span></span>
<span><span class="fu"><a href="../reference/plot_pcoa.html">plot_pcoa</a></span><span class="op">(</span><span class="va">pcoa</span>, num_components <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://github.com/mrc-ide/MIPanalyzer/raw/master/vignettes/PCOA_scatter.png" style="width: 80%"></p>
<p>As before, we find little evidence of clustering, although there are
a few interesting outliers. This implies that most samples have roughly
the same genetic distance to all others, and so there is no strong
signal of clustering of allele frequencies.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Bob Verity.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
