<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ MIPanalyzer</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MIPanalyzer</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/filtering_data.html">Filtering data</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="MIPanalyzer.html">MIPanalyzer</a></code> <code><a href="MIPanalyzer.html">MIPanalyzer-package</a></code> </p>
        </td>
        <td><p>MIPanalyzer</p></td>
      </tr><tr><td>
          <p><code><a href="Pf_chrom_lengths.html">Pf_chrom_lengths()</a></code> </p>
        </td>
        <td><p>Get dataframe of P.falciparum chromosome lengths</p></td>
      </tr><tr><td>
          <p><code><a href="check_MIPanalyzer_loaded.html">check_MIPanalyzer_loaded()</a></code> </p>
        </td>
        <td><p>Check that MIPanalyzer package has loaded successfully</p></td>
      </tr><tr><td>
          <p><code><a href="explore_filter_coverage_loci.html">explore_filter_coverage_loci()</a></code> </p>
        </td>
        <td><p>Explore locus coverage prior to filtering</p></td>
      </tr><tr><td>
          <p><code><a href="explore_filter_coverage_samples.html">explore_filter_coverage_samples()</a></code> </p>
        </td>
        <td><p>Explore sample coverage prior to filtering</p></td>
      </tr><tr><td>
          <p><code><a href="filter_counts.html">filter_counts()</a></code> </p>
        </td>
        <td><p>Filter alleles based on raw counts</p></td>
      </tr><tr><td>
          <p><code><a href="filter_coverage_loci.html">filter_coverage_loci()</a></code> </p>
        </td>
        <td><p>Filter loci based on coverage</p></td>
      </tr><tr><td>
          <p><code><a href="filter_coverage_samples.html">filter_coverage_samples()</a></code> </p>
        </td>
        <td><p>Filter samples based on coverage</p></td>
      </tr><tr><td>
          <p><code><a href="filter_loci.html">filter_loci()</a></code> </p>
        </td>
        <td><p>Filter out some loci</p></td>
      </tr><tr><td>
          <p><code><a href="filter_loci_invariant.html">filter_loci_invariant()</a></code> </p>
        </td>
        <td><p>Filter loci to drop invariant sites</p></td>
      </tr><tr><td>
          <p><code><a href="filter_overcounts.html">filter_overcounts()</a></code> </p>
        </td>
        <td><p>Filter out over-counts</p></td>
      </tr><tr><td>
          <p><code><a href="filter_samples.html">filter_samples()</a></code> </p>
        </td>
        <td><p>Filter out some samples</p></td>
      </tr><tr><td>
          <p><code><a href="filter_wsaf.html">filter_wsaf()</a></code> </p>
        </td>
        <td><p>Filter alleles based on within-sample allele frequencies</p></td>
      </tr><tr><td>
          <p><code><a href="get_IBS_distance.html">get_IBS_distance()</a></code> </p>
        </td>
        <td><p>Get identity by state (IBS) distance</p></td>
      </tr><tr><td>
          <p><code><a href="get_IB_mixture.html">get_IB_mixture()</a></code> </p>
        </td>
        <td><p>Get identity by mixture</p></td>
      </tr><tr><td>
          <p><code><a href="get_genomic_distance.html">get_genomic_distance()</a></code> </p>
        </td>
        <td><p>Get genomic distance between samples</p></td>
      </tr><tr><td>
          <p><code><a href="get_spatial_distance.html">get_spatial_distance()</a></code> </p>
        </td>
        <td><p>Get great circle distance between spatial points</p></td>
      </tr><tr><td>
          <p><code><a href="get_wsaf.html">get_wsaf()</a></code> </p>
        </td>
        <td><p>Get within-sample allele frequencies</p></td>
      </tr><tr><td>
          <p><code><a href="inbreeding_mle.html">inbreeding_mle()</a></code> </p>
        </td>
        <td><p>Estimate pairwise inbreeding coefficient F by maximum likelihood</p></td>
      </tr><tr><td>
          <p><code><a href="is.mipanalyzer_biallelic.html">is.mipanalyzer_biallelic()</a></code> </p>
        </td>
        <td><p>Determine if object is of class mipanalyzer_biallelic</p></td>
      </tr><tr><td>
          <p><code><a href="is.mipanalyzer_multiallelic.html">is.mipanalyzer_multiallelic()</a></code> </p>
        </td>
        <td><p>Determine if object is of class mipanalyzer_multiallelic</p></td>
      </tr><tr><td>
          <p><code><a href="lonlat_to_bearing.html">lonlat_to_bearing()</a></code> </p>
        </td>
        <td><p>Calculate great circle distance and bearing between coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="mipanalyzer_biallelic_to_vcfR.html">mipanalyzer_biallelic_to_vcfR()</a></code> </p>
        </td>
        <td><p>Converts a MIPanalyzer biallelic object to vcfR</p></td>
      </tr><tr><td>
          <p><code><a href="mipanalyzer_file.html">mipanalyzer_file()</a></code> </p>
        </td>
        <td><p>Load system file</p></td>
      </tr><tr><td>
          <p><code><a href="pca_wsaf.html">pca_wsaf()</a></code> </p>
        </td>
        <td><p>PCA of within-sample allele frequencies</p></td>
      </tr><tr><td>
          <p><code><a href="pcoa_genomic_distance.html">pcoa_genomic_distance()</a></code> </p>
        </td>
        <td><p>PCoA of genomic distances between samples</p></td>
      </tr><tr><td>
          <p><code><a href="plot_coverage.html">plot_coverage()</a></code> </p>
        </td>
        <td><p>Plot coverage matrix</p></td>
      </tr><tr><td>
          <p><code><a href="plot_map.html">plot_map()</a></code> </p>
        </td>
        <td><p>Produce ggplot map</p></td>
      </tr><tr><td>
          <p><code><a href="plot_pca.html">plot_pca()</a></code> </p>
        </td>
        <td><p>Plot PCA</p></td>
      </tr><tr><td>
          <p><code><a href="plot_pca_contribution.html">plot_pca_contribution()</a></code> </p>
        </td>
        <td><p>Plot PCA contribution of each variable</p></td>
      </tr><tr><td>
          <p><code><a href="plot_pca_variance.html">plot_pca_variance()</a></code> </p>
        </td>
        <td><p>Plot variance explained by PCA components</p></td>
      </tr><tr><td>
          <p><code><a href="plot_pcoa.html">plot_pcoa()</a></code> </p>
        </td>
        <td><p>Plot PCoA</p></td>
      </tr><tr><td>
          <p><code><a href="print.mipanalyzer_biallelic.html">print(<i>&lt;mipanalyzer_biallelic&gt;</i>)</a></code> </p>
        </td>
        <td><p>Custom print function for class mipanalyzer_biallelic</p></td>
      </tr><tr><td>
          <p><code><a href="print.mipanalyzer_multiallelic.html">print(<i>&lt;mipanalyzer_multiallelic&gt;</i>)</a></code> </p>
        </td>
        <td><p>Custom print function for class mipanalyzer_multiallelic</p></td>
      </tr><tr><td>
          <p><code><a href="print_full.html">print_full()</a></code> </p>
        </td>
        <td><p>Ordinary print function for unclassed object</p></td>
      </tr><tr><td>
          <p><code><a href="rbetabinom.html">rbetabinom()</a></code> </p>
        </td>
        <td><p>Draw from Beta-binomial distribution</p></td>
      </tr><tr><td>
          <p><code><a href="rdirichlet.html">rdirichlet()</a></code> </p>
        </td>
        <td><p>Draw from Dirichlet distribution</p></td>
      </tr><tr><td>
          <p><code><a href="sim_biallelic.html">sim_biallelic()</a></code> </p>
        </td>
        <td><p>Simulate biallelic data</p></td>
      </tr><tr><td>
          <p><code><a href="summary.mipanalyzer_biallelic.html">summary(<i>&lt;mipanalyzer_biallelic&gt;</i>)</a></code> </p>
        </td>
        <td><p>Print summary for class mipanalyzer_biallelic</p></td>
      </tr><tr><td>
          <p><code><a href="summary.mipanalyzer_multiallelic.html">summary(<i>&lt;mipanalyzer_multiallelic&gt;</i>)</a></code> </p>
        </td>
        <td><p>Print summary for class mipanalyzer_multiallelic</p></td>
      </tr><tr><td>
          <p><code><a href="vcf_to_mipanalyzer_biallelic.html">vcf_to_mipanalyzer_biallelic()</a></code> </p>
        </td>
        <td><p>Convert vcf to biallelic mipanalyzer data class</p></td>
      </tr><tr><td>
          <p><code><a href="vcf_to_mipanalyzer_multiallelic.html">vcf_to_mipanalyzer_multiallelic()</a></code> </p>
        </td>
        <td><p>Convert vcf to multiallelic mipanalyzer data class</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Bob Verity.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

